File: sensor-ultrasonic-distance/ultrasonic-distance.c
/*
 * 
 * This is not finished, use the python version instead.
 ***********************************************************************
 */

#include <stdio.h>
#include <wiringPi.h>

// LED Pin - wiringPi pin 0 is BCM_GPIO 17.

#define	PTRIG 17
#define PECHO 18

int timerStart = 0;

void interruptHandler(void)
{
	//printf("interrupt!S"); fflush (stdout);
	int now = micros();
	int delay = now - timerStart;
	printf("%f\n", delay / 58.0);	fflush (stdout); 
	
	//cm = pulseIn(PECHO, HIGH) / 58.0; //Echo time convert to cm
	//cm = (int(cm * 100.0)) / 100.0; //Keep two decimal places
}


int main(void)
{
	printf("Raspberry Pi distance measure\n");

	wiringPiSetupGpio();
	pinMode(PTRIG, OUTPUT);
	//pinMode(PECHO, INPUT);
	//pullUpDnControl(PECHO,PUD_DOWN);
	wiringPiISR(PECHO, INT_EDGE_RISING, &interruptHandler);
  	printf("callback registered\n");



	
	
	for(;;)
	{
	digitalWrite(PTRIG, LOW); //Send a voltage pulse to TrigPin
	delayMicroseconds(2);
	digitalWrite(PTRIG, HIGH);
	delayMicroseconds(10);
	digitalWrite(PTRIG, LOW);
	timerStart = micros();
	
	delay(200);
		
	}


  return 0 ;
}

File: sensor-ultrasonic-distance/ultrasonic-distance.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import sys


import RPi.GPIO as GPIO
import time


ptrig = 26
pecho = 27

def setup():
	GPIO.setwarnings(False)
	GPIO.setmode(GPIO.BCM)
	GPIO.setup(pecho, GPIO.IN, pull_up_down=GPIO.PUD_DOWN)
	GPIO.setup(ptrig, GPIO.OUT)
	GPIO.output(ptrig, 0)
		


def getValue():
	GPIO.output(ptrig, 0)
	time.sleep(0.1)
	GPIO.output(ptrig, 1)
	time.sleep(0.00001)
	GPIO.output(ptrig, 0)
	
	
	while(0 == GPIO.input(pecho)):
		start = time.time()
	while(1 == GPIO.input(pecho)):
		end = time.time()
		

	delay = (end - start) * 1000 * 1000
	time.sleep(0.1)
	distance = (delay / 58.0)
	if 2 < distance < 400: # working interval
		print("distance: %0.1f cm" % distance)


setup()
while(True):
	getValue()
print('done.')

